### PR TITLE
Fix deletion query for incremental ingestion backend

### LINKED ISSUE #92: [Incremental Ingestion Backend] Incorrect query used for deletion
We found that the query we used to identify records to delete was incorrect. After ingestion, we were getting a query error. 

We fixed the problem in the closed source version that we need to bring into the open source.

You need to change `CHANGE_ENTITY_NAME`.

```ts
        const removed: { entity: string; ref: string }[] = await tx('final_entities')
          .select(tx.ref('final_entity').as('entity'), tx.ref('refresh_state.entity_ref').as('ref'))
          .join(tx.raw('refresh_state ON refresh_state.entity_id = final_entities.entity_id'))
          .whereRaw(`((final_entity::json #>> '{metadata, annotations, CHANGE_ENTITY_NAME}')) = ?`, [
            provider.getProviderName(),
          ])
          .whereNotIn(
            'entity_ref',
            tx('ingestion.ingestion_marks')
              .join(
                'ingestion.ingestion_mark_entities',
                'ingestion.ingestion_marks.id',
                'ingestion.ingestion_mark_entities.ingestion_mark_id',
              )
              .select('ingestion.ingestion_mark_entities.ref')
              .where('ingestion.ingestion_marks.ingestion_id', id),
          );
```

### LINKED ISSUE #93: [Incremental Ingestion Backend] Must JSON.parse response from remove entities query
In our client's project, we found that after applying solution in #92, removing entities was causing Malformed Entities to be committed. We traced it to remove query returning strings instead of JSON objects for entities. We added the following code to remove logic. 

```ts
return removed.map(entity => {
  return { entity: JSON.parse(entity.entity) };
});
```

We need to add the same logic to our open source version of Incremental Ingestion Backend.

### LINKED ISSUE #94: [Incremental Ingestion Backend] We shouldn't commit when there are no entities
Incremental Ingestion Backend should only attempt to insert marks when there are entities to mark. In our close source version, we added the following code

```ts
    if (entities.length > 0) {
      await tx('ingestion.ingestion_mark_entities').insert(
        entities.map(entity => ({
          id: v4(),
          ingestion_mark_id: markId,
          ref: stringifyEntityRef(entity.entity),
        })),
      );
    }
```

### PR DESCRIPTION
**Pull request by wKich**
## Motivation

Incremental ingestion backend package has a few issues which were fixed in closed source version.
Fixes #92
Fixes #93
Fixes #94

## Approach

Sync fixes from closed source version

### CODE REVIEW

File: plugins/incremental-ingestion-backend/src/iteration-db.ts [hunk 1]
```diff
@@ -204,7 +204,7 @@ export async function createIterationDB(options: IterationDBOptions): Promise<It
       // eslint-disable-next-line no-constant-condition
       while (true) {
         done = next.done;
-        await mark(tx, id, sequence, next.entities, next.done, next.cursor);
+        await mark(tx, id, sequence, next.entities, next.done, JSON.stringify(next.cursor));
```
Comment (taras) [inline plugins/incremental-ingestion-backend/src/iteration-db.ts:205]: Can you tell me more about this change?

**Review by taras — APPROVED** [review]
👍

File: plugins/incremental-ingestion-backend/src/iteration-db.ts [hunk 2]
```diff
@@ -267,23 +267,25 @@ export async function createIterationDB(options: IterationDBOptions): Promise<It
...
-          .whereNotIn(
-            'ref',
-            tx('ingestion.ingestion_marks')
-              .join(
-                'ingestion.ingestion_mark_entities',
-                'ingestion.ingestion_marks.id',
-                'ingestion.ingestion_mark_entities.ingestion_mark_id',
-              )
-              .select('ingestion.ingestion_mark_entities.ref'),
+        removed = (
```
Comment (cowboyd) [inline plugins/incremental-ingestion-backend/src/iteration-db.ts]: Let's pull this into a function. This is a big enough block that it makes tracking the mutability non-trivial.

**Review by cowboyd — CHANGES_REQUESTED** [review]
This looks good. 

It's more apparent with every change that we need to settle on a testing strategy for this deletion logic. Not now, but every time we change it, it's a risk. According to our conversation yesterday, it sounds like we'll need to add deletion to graphgen.

**Comment by frontsidejack** [comment]
📣 NOTIFICATION
You are receiving this message because we did not publish any packages.
<p align='right'><em>Generated by <a href='https://github.com/thefrontside/actions/tree/main/publish-pr-preview'>@thefrontside/actions Frontside</a><em></p>

**Comment by wKich** [comment]
@cowboyd Oh, I'm sorry, I've merged branch without your approval

**Comment by cowboyd** [comment]
It looks like we didn't have branch protection enabled.

**Pull request merged**